Userland/Libraries/LibJS/Runtime/PromiseConstructor.cpp

```diff
@@ -80,7 +196,32 @@ Value PromiseConstructor::construct(FunctionObject& new_target)
 // 27.2.4.1 Promise.all ( iterable ), https://tc39.es/ecma262/#sec-promise.all
 JS_DEFINE_NATIVE_FUNCTION(PromiseConstructor::all)
 {
-    TODO();
+    auto* constructor = vm.this_value(global_object).to_object(global_object);
+    if (!constructor)
+        return {};
+
+    auto promise_capability = new_promise_capability(global_object, constructor);
+    if (vm.exception())
+        return {};
+
+    auto promise_resolve = get_promise_resolve(global_object, constructor);
+    if (auto abrupt = if_abrupt_reject_promise(global_object, promise_resolve, promise_capability); abrupt.has_value())
+        return abrupt.value();
+
+    auto iterator_record = get_iterator(global_object, vm.argument(0));
+    if (auto abrupt = if_abrupt_reject_promise(global_object, iterator_record, promise_capability); abrupt.has_value())
+        return abrupt.value();
+
+    auto result = perform_promise_all(global_object, *iterator_record, constructor, promise_capability, promise_resolve);
+    if (vm.exception()) {
+        if (!iterator_record_is_complete(global_object, *iterator_record))
+            iterator_close(*iterator_record);
+
+        if (auto abrupt = if_abrupt_reject_promise(global_object, result, promise_capability); abrupt.has_value())
+            return abrupt.value();
```

Comment:
We're already in an `if (vm.exception())` block, so this should always return a value - can we do this instead?

```suggestion
        auto abrupt = if_abrupt_reject_promise(global_object, result, promise_capability);
        return abrupt.value();
```